File: tools/pauses/endings.py
"""Find where the hiss starts in each of a voice's lines ending on a nasal (FR-555).

Reads one JSON request on standard input: every setting, then the WAV files of one
voice's lines. Writes a JSON list on standard output answering each file in order with
the sample its hiss starts at; null where the line ends on no burst. The Go half of the
pauses tool owns every setting, the lines, the fade and the file it writes, so this half
only reads the sound.

A burst is a run of frames, counted from the line's first sample, each louder than the
request's loudness with at least its share of energy above its frequency, read through
a Hann window, whose last frame ends no more than its span before the end of the line's
last frame louder than that loudness. The hiss starts at the first of the shorter hiss
frames read the same way, from one frame before the run's first frame up to it; at the
run's first sample where none of them is a burst frame.
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass
from typing import Any

import numpy as np
import numpy.typing as npt
import soundfile

# Decibels are twenty times the base-ten logarithm of an amplitude ratio.
DECIBELS_PER_DECADE = 20

Samples = npt.NDArray[np.float32]


@dataclass(frozen=True, slots=True)
class Settings:
    """The request's settings.

    A frame and a hiss frame in samples, the frequency in hertz a burst frame holds its
    share of energy above, the loudness in decibels a frame must exceed, that share and
    the most samples a burst may end before the line's last loud frame ends.
    """

    frame: int
    hiss_frame: int
    high_hz: float
    loud_db: float
    share: float
    within: int

    @classmethod
    def read(cls, request: dict[str, Any]) -> Settings:
        """Take every setting from the request; a missing one fails the run."""
        return cls(
            frame=int(request["frame"]),
            hiss_frame=int(request["hiss_frame"]),
            high_hz=float(request["high_hz"]),
            loud_db=float(request["loud_db"]),
            share=float(request["share"]),
            within=int(request["within"]),
        )


@dataclass(frozen=True, slots=True)
class Reading:
    """What one frame is: louder than the loudness; a burst frame besides."""

    loud: bool
    burst: bool


def read_frame(frame: Samples, rate: int, settings: Settings) -> Reading:
    """Read one frame: loud where it beats the loudness, a burst where its share is high too."""
    values = frame.astype(np.float64)
    rms = float(np.sqrt(np.mean(values**2)))
    if rms <= 0 or DECIBELS_PER_DECADE * np.log10(rms) <= settings.loud_db:
        return Reading(loud=False, burst=False)
    power = np.abs(np.fft.rfft(values * np.hanning(len(values)))) ** 2
    frequencies = np.fft.rfftfreq(len(values), 1 / rate)
    high = float(power[frequencies >= settings.high_hz].sum())
    return Reading(loud=True, burst=high >= settings.share * float(power.sum()))


def run_start(samples: Samples, rate: int, settings: Settings) -> int | None:
    """Answer the first sample of the last run of burst frames; None without a burst."""
    start: int | None = None
    burst_end: int | None = None
    loud_end = 0
    in_run = False
    for first in range(0, len(samples) - settings.frame + 1, settings.frame):
        reading = read_frame(samples[first : first + settings.frame], rate, settings)
        if not reading.loud:
            in_run = False
            continue
        loud_end = first + settings.frame
        if reading.burst:
            if not in_run:
                start = first
            burst_end = first + settings.frame
        in_run = reading.burst
    if burst_end is None or loud_end - burst_end > settings.within:
        return None
    return start


def hiss_start(samples: Samples, rate: int, settings: Settings) -> int | None:
    """Answer the sample the hiss starts at; None without a burst."""
    run = run_start(samples, rate, settings)
    if run is None:
        return None
    for first in range(max(0, run - settings.frame), run, settings.hiss_frame):
        hiss_frame = samples[first : first + settings.hiss_frame]
        if read_frame(hiss_frame, rate, settings).burst:
            return first
    return run


def answer(path: str, settings: Settings) -> dict[str, int | None]:
    """Answer where one file's hiss starts; None where it ends on no burst."""
    samples, rate = soundfile.read(path, dtype="float32")
    return {"hiss": hiss_start(samples, rate, settings)}


def main() -> None:
    """Answer every file the request names, in order."""
    sys.stdin.reconfigure(encoding="utf-8")
    sys.stdout.reconfigure(encoding="utf-8")
    request = json.load(sys.stdin)
    settings = Settings.read(request)
    json.dump([answer(path, settings) for path in request["files"]], sys.stdout)


if __name__ == "__main__":
    main()

File: tools/pauses/endings.go
package main

import (
	"context"
	"time"

	"github.com/oernster/bridge-talk/internal/domain/ending"
	"github.com/oernster/bridge-talk/internal/domain/machinevoice"
	"github.com/oernster/bridge-talk/internal/domain/measured"
	"github.com/oernster/bridge-talk/internal/domain/script"
)

// Every setting of the fade after a final nasal has its one home here. endings.py holds none: it reads
// each from the request (FR-555, FR-556).
const (
	// fadeLength is how long a fade lasts, ending where the hiss starts, chosen by ear (FR-556).
	fadeLength = 30 * time.Millisecond
	// burstFrame is the length of each frame a burst is read in (FR-555).
	burstFrame = 10 * time.Millisecond
	// hissFrame is the length of each frame the start of the hiss is read in, over the burst frame
	// before a burst (FR-555).
	hissFrame = 500 * time.Microsecond
	// burstWithin is the longest a burst's last frame may end before the line's last loud frame ends
	// (FR-555).
	burstWithin = 50 * time.Millisecond
	// burstHighHz is the frequency above which a burst frame holds its share of energy (FR-555).
	burstHighHz = 3000
	// burstLoudDB is the loudness a frame must exceed to be read at all, in decibels (FR-555).
	burstLoudDB = -50
	// burstShare is the least share of a frame's energy above burstHighHz that makes it a burst frame
	// (FR-555).
	burstShare = 0.4
)

// fadedWord is what the tool prints of the lines given a fade (FR-555).
const fadedWord = "faded"

// endingRequest is what endings.py is handed for one voice: every setting, then the voice's lines
// ending on a nasal written as WAV files in order. Frames and spans are counted in samples.
type endingRequest struct {
	// Frame is the length of a frame in samples.
	Frame int `json:"frame"`
	// HissFrame is the length in samples of each frame the start of the hiss is read in.
	HissFrame int `json:"hiss_frame"`
	// HighHz is the frequency above which a burst frame holds its share of energy.
	HighHz float64 `json:"high_hz"`
	// LoudDB is the loudness in decibels a frame must exceed to be read.
	LoudDB float64 `json:"loud_db"`
	// Share is the least share of a frame's energy above HighHz that makes it a burst frame.
	Share float64 `json:"share"`
	// Within is the most samples a burst may end before the line's last loud frame ends.
	Within int `json:"within"`
	// Files is each line's WAV file, in the order the script lists the lines.
	Files []string `json:"files"`
}

// ended is endings.py's answer for one file: the sample the hiss starts at, nil where the line ends on
// no burst.
type ended struct {
	// Hiss is the sample the hiss starts at.
	Hiss *int `json:"hiss"`
}

// endingFinder finds where each of a voice's lines written as WAV files ends on a burst, answering in
// the same order.
type endingFinder interface {
	FindEndings(asked endingRequest) ([]ended, error)
}

// endingRequestFor asks where a voice's lines end on a burst.
func endingRequestFor(files []string) endingRequest {
	return endingRequest{
		Frame: samplesIn(burstFrame), HissFrame: samplesIn(hissFrame), HighHz: burstHighHz, LoudDB: burstLoudDB, Share: burstShare,
		Within: samplesIn(burstWithin), Files: files,
	}
}

// endings finds the endings of each voice over the lines ending on a nasal in its accent, printing
// each voice then the total, then builds the book (FR-555).
func (f finding) endings(ctx context.Context, voiced script.Voiced, voices []machinevoice.Voice) (ending.Book, error) {
	faded, model, err := eachVoice(f, voices, fadedWord, func(voice machinevoice.Voice) (measuredVoice[ending.Voice], error) {
		return f.endingsOf(ctx, voice, voiced.EndingOnNasal(voice.Accent()))
	})
	if err != nil {
		return ending.Book{}, err
	}
	return ending.NewBook(samplesIn(fadeLength), model, faded)
}

// endingsOf makes each of a voice's lines ending on a nasal for the ending finder, building the voice's
// endings from what it answers and printing the voice's faded lines. Each fade starts its length before
// the sample the hiss starts at, so it ends where the hiss starts (FR-556).
func (f finding) endingsOf(ctx context.Context, voice machinevoice.Voice, lines []script.SavedLine) (measuredVoice[ending.Voice], error) {
	var answers []ended
	made, err := f.makeFor(ctx, voice, lines, func(files []string) (int, error) {
		var err error
		answers, err = f.ends.FindEndings(endingRequestFor(files))
		return len(answers), err
	})
	if err != nil {
		return measuredVoice[ending.Voice]{}, err
	}
	entries := make([]ending.Entry, 0, len(lines))
	var listed []string
	for at, line := range lines {
		entry := ending.Entry{Cue: line.Cue, Index: line.Index, Sounds: line.Sounds, Digest: made.digests[at]}
		if hiss := answers[at].Hiss; hiss != nil {
			start := *hiss - samplesIn(fadeLength)
			if err := within(start, made.lengths[at], voice, line); err != nil {
				return measuredVoice[ending.Voice]{}, err
			}
			entry.Sample = start
			listed = append(listed, measured.LineName(voice.ID(), line.Cue, line.Index))
		}
		entries = append(entries, entry)
	}
	f.list(voice, fadedWord, listed, len(entries))
	return measuredVoice[ending.Voice]{
		entries: ending.Voice{Style: made.from.Style, Entries: entries}, model: made.from.Model, lines: len(entries), marked: len(listed),
	}, nil
}
